common/src/main/java/com/tc/net/protocol/transport/ClientConnectionEstablisher.java

```diff
@@ -372,111 +375,93 @@ private void quitReconnectAttempts() {
-        isInterrupted = true;
-      } finally {
-        Util.selfInterruptIfNeeded(isInterrupted);
+        reconnect.getCurrentTask().get();
+      } catch (CancellationException cancelled) {
+        // do nothing task is complete
       }
-      return false; //  even if interrupted, don't reschedule
     }
-
-    // for testing only
-    private boolean waitForThreadToComplete() {
-      return waitForThread(getConnectionThread(), false);
+  }
+  // for testing only
+  void interruptReconnect() {
+    AsyncReconnect reconnect = getReconnectHandler();
+    if (reconnect != null) {
+      reconnect.getCurrentTask().cancel(true);
     }
-
-    private void awaitTermination(boolean mayInterruptIfRunning) {
-      Assert.assertTrue(isStopped());
-      waitForThread(getConnectionThread(), mayInterruptIfRunning);
+  }
+  
+  boolean isReconnecting() {
+    AsyncReconnect reconnect = getReconnectHandler();
+    if (reconnect != null) {
+      return !reconnect.getCurrentTask().isDone();
+    } else {
+      return false;
     }
-
-    public synchronized boolean isStopped() {
-      return stopped;
+  }
+  
+  private class AsyncReconnect {
+    private final ExecutorService connectionEstablisher = new ThreadPoolExecutor(0, 1, 5, TimeUnit.SECONDS, new SynchronousQueue<>(), (Runnable r) -> {
+      Thread t = new Thread(r, RECONNECT_THREAD_NAME + "-" + serverAddresses.toString() + "-" + transport.getConnectionID());
+      t.setDaemon(true);
+      return t;
+    });
+    private volatile Future<?> currentTask = CompletableFuture.completedFuture(null);
+
+    public boolean isStopped() {
+      return connectionEstablisher.isShutdown();
     }
     
-    public synchronized void stop() {
+    public void stop() {
       LOGGER.debug("Connection establisher stopping for connection {} to {}", transport.getConnectionID(), serverAddresses);
-      stopped = true;
-      this.notifyAll();
+      connectionEstablisher.shutdown();
+      currentTask.cancel(true);
     }
     
-    private synchronized Thread getConnectionThread() {
-      return connectionEstablisherThread;
+    public Future<?> getCurrentTask() {
+      return currentTask;
     }
 
     public boolean putConnectionRequest(ConnectionRequest request) {
-      if (!isStopped() && waitForThread(getConnectionThread(), true)) {
-        if (!transport.isConnected()) {
-          startThreadIfNecessary(request);
+      if (isStopped()) {
+        LOGGER.info("connect request ignored, reconnect has been shut down");
+        return false;
+      } else if (!currentTask.isDone()) {
+        //  if the current task is not done, this addtional request is likely 
+        //  due to a failed reconnect, let the original just continue on
+        LOGGER.info("connect request ignored, already reconnecting");
+        return false;
+      } else if (!transport.isConnected()) {
+        try {
+          currentTask = connectionEstablisher.submit(()->execute(request));
```

Comment:
Isn't this going to block because of the `SynchronousQueue` if two of these calls race?

Reply:
no.  I used that queue as kind of a guard.  Since we are making sure the previous task is complete before adding a new one, the queue should be empty.

Reply:
Right but could there be two calls to this that both see `currentTask.isDone()` as `true and `transport.isConnected()` as `false`, and then they both end up submitting to the executor?

Reply:
Trying to mimic what was there before.  Only one task ever in flight at one time.   The transport.isConnected check should be an assert really.  the code should never get to that point while being connected.